presto-main/src/main/java/com/facebook/presto/sql/analyzer/Windows.java

```diff
@@ -0,0 +1,136 @@
+import com.facebook.presto.sql.tree.WindowSpecification;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.INVALID_WINDOW_SPECIFICATION;
+import static com.google.common.base.Preconditions.checkArgument;
+
+final class Windows
+{
+    private Windows()
+    {
+    }
+
+    static Map<String, WindowSpecification> createWindowSpecificationMap(Node node, List<WindowDefinition> windowDefinitions, Function<String, WindowSpecification> windowSpecificationLookup)
+    {
+        Map<String, WindowSpecification> windowSpecifications = new HashMap<>();
+        for (WindowDefinition def : windowDefinitions) {
+            if (windowSpecifications.containsKey(def.getName())) {
+                throw new SemanticException(INVALID_WINDOW_SPECIFICATION, node, "Duplicate window definition '%s'", def.getName());
+            }
+            WindowSpecification spec = def.getSpecification();
+            if (spec.getExistingName().isPresent()) {
+                WindowSpecification existing = windowSpecifications.get(spec.getExistingName().get());
+                if (existing == null) {
+                    existing = windowSpecificationLookup.apply(spec.getExistingName().get());
+                    if (existing == null) {
+                        throw new SemanticException(INVALID_WINDOW_SPECIFICATION, node, "Existing window definition '%s' not found for '%s'", spec.getExistingName().get(), def.getName());
+                    }
+                }
+                spec = resolveWindowReference(node, spec, existing);
+            }
+            windowSpecifications.put(def.getName(), spec);
+        }
+        return ImmutableMap.copyOf(windowSpecifications);
+    }
+
+    private static WindowSpecification resolveWindowReference(Node node, WindowSpecification referrer, WindowSpecification referent)
+    {
+        checkArgument(referrer.getExistingName().isPresent(), "Window specification does not contain reference");
+
+        // These rules are arcane and inconsistent, see section 7.11 of the SQL 2008 spec for details.
+        // Alternatively, look at transformWindowDefinitions in backend/parser/parse_clause.c in pg.
+
+        if (referent.getExistingName().isPresent()) {
+            throw new SemanticException(INVALID_WINDOW_SPECIFICATION, node, "Cannot chain window specification references");
+        }
+
+        if (!referrer.getPartitionBy().isEmpty()) {
+            throw new SemanticException(INVALID_WINDOW_SPECIFICATION, node, "Referrer window specification must not contain partition clauses");
+        }
+        List<Expression> partitionBy = ImmutableList.copyOf(referent.getPartitionBy());
+
+        List<SortItem> orderBy;
+        if (referrer.getOrderBy().isEmpty()) {
```

Comment:
My interpretation is based on 7.11 Syntax Rules 10.d and 7.11 General Rules 1.b.i.5.A-C, by the way.